File: frontend/src/app/sites/qwirkle/qwirkle.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {Observable} from "rxjs";
import {GameState} from "./dto/game/GameState";
import {Move} from "./dto/move/Move";
import {Tile} from "./dto/tile/Tile";
import {MoveGroup} from "./dto/move/MoveGroup";
import {SelectionInfo} from "./dto/SelectionInfo";
import {GameSessionService} from "../all/services/http/game-session.service";
import {GameConfig} from "./dto/game/GameConfig";
import {environment} from "../../../environments/environment";
import {NotificationService} from "../../main/core/services/notification.service";

@Injectable({
    providedIn: 'root'
})
export class QwirkleService extends GameSessionService<GameConfig, GameState> {
    protected readonly baseUri: string = environment.httpApiUrl + '/qwirkle';

    constructor(httpClient: HttpClient, notification: NotificationService) {
        super(httpClient, notification);
    }

    clearHand(): Observable<GameState> {
        return this.httpClient.post<GameState>(this.baseUri + '/hand/clear', {});
    }

    getSelectionInfo(selected: Tile[], fromStack: boolean = false): Observable<SelectionInfo> {
        return this.httpClient.post<SelectionInfo>(this.baseUri + '/selection', selected, {
            params: {
                fromStack: fromStack
            }
        });
    }

    getBestMoves(maxMoves: number = 1): Observable<MoveGroup[]> {
        return this.httpClient.get<MoveGroup[]>(this.baseUri + '/solve', {
            params: {maxMoves: maxMoves.toString()}
        });
    }

    drawTile(tile: Tile): Observable<GameState> {
        return this.httpClient.post<GameState>(this.baseUri + '/stack/draw', tile);
    }

    makeMove(move: Move, fromStack: boolean = false): Observable<GameState> {
        return this.httpClient.post<GameState>(this.baseUri + '/move', move, {
            params: {
                fromStack: fromStack
            }
        });
    }

    undoMove() : Observable<GameState> {
        return this.httpClient.post<GameState>(this.baseUri + '/undo', {});
    }

    uploadImage(image: File): Observable<Blob> {
        const formData = new FormData();
        formData.append('file', image);

        return this.httpClient.post<Blob>(this.baseUri + '/image/upload', formData, {
            responseType: 'blob' as 'json'
        }) as Observable<Blob>;
    }

    confirmImage(): Observable<void> {
        return this.httpClient.post<void>(this.baseUri + '/image/confirm', {});
    }
}

File: backend/src/main/java/net/zorphy/backend/site/qwirkle/service/QwirkleService.java
package net.zorphy.backend.site.qwirkle.service;

import net.zorphy.backend.site.core.http.service.GameSessionService;
import net.zorphy.backend.site.core.http.service.SavableService;
import net.zorphy.backend.site.core.http.dto.ResultState;
import net.zorphy.backend.site.qwirkle.dto.SelectionInfo;
import net.zorphy.backend.site.qwirkle.dto.game.GameConfig;
import net.zorphy.backend.site.qwirkle.dto.game.GameState;
import net.zorphy.backend.site.qwirkle.dto.move.Move;
import net.zorphy.backend.site.qwirkle.dto.move.MoveGroup;
import net.zorphy.backend.site.qwirkle.dto.tile.Tile;

import java.util.List;

public interface QwirkleService extends GameSessionService<GameConfig, GameState>, SavableService<GameState, ResultState> {
    /**
     * Retrieves the {@code N} best moves for the given {@code gameState} where {@code N} is {@code maxMoves}.
     */
    List<MoveGroup> getBestMoves(GameState gameState, int maxMoves);

    /**
     * Draws the given {@code tile} from the stack in the given {@code gameState}.
     */
    GameState drawTile(GameState gameState, Tile tile);

    /**
     * Retrieves the selection information for the given {@code tiles} in the given {@code gameState}.
     */
    SelectionInfo getSelectionInfo(GameState gameState, List<Tile> tiles, boolean fromStack);

    /**
     * Clears the hand in the given {@code gameState}.
     */
    GameState clearHand(GameState gameState);

    /**
     * Makes the given {@code move} in the given {@code gameState}.
     */
    GameState makeMove(GameState gameState, Move move, boolean fromStack);

    GameState undoMove(GameState gameState);

    byte[] uploadImage(byte[] file);
}
